### ISSUE TITLE
Unexpected behavior from differentiation on vector norm()

### ISSUE THREAD

**Issue by kracon7** [taichi-dev/taichi#4332]
**Describe the bug**
Unexpected behavior from differentiation on vector norm(). Got gradient of NaN

**To Reproduce**
```py
import numpy as np
import taichi as ti

ti.init()

DTYPE = ti.f64

@ti.data_oriented
class ExampleClass:
    def __init__(self):
      
        k = 2
        
        # mass of each circle
        self.x = ti.field(DTYPE, k, needs_grad=True)
        self.x.from_numpy(np.ones(k))
      
        # pos, vel and force of each circle
        self.y = ti.Vector.field(2, DTYPE, shape=(2, k), needs_grad=True)
 
        self.z = ti.Vector.field(2, DTYPE, shape=(), needs_grad=True)

        self.n = ti.field(DTYPE, (), needs_grad=True)

    @ti.kernel
    def test(self):
        self.z[None] += self.x[0] * self.y[0, 0]

        self.n[None] += self.x[0] * (self.y[0, 0] - self.z[None]).norm()**2
        
        # # change norm() to norm(1e-5) and the gradient is no longer NaN
        # self.n[None] += self.x[0] * (self.y[0, 0] - self.z[None]).norm(1e-5)**2
            
@ti.data_oriented
class Loss():
    """docstring for Loss"""
    def __init__(self, engines):
        self.loss = ti.field(DTYPE, shape=(), needs_grad=True)
        self.engines = engines

    @ti.kernel
    def clear_loss(self):
        self.loss[None] = 0

    @ti.kernel
    def compute_loss(self):
        self.loss[None] = self.engines[0].y[1, 0][0] - self.engines[1].y[1, 0][0]
                              

sim1 = ExampleClass()
sim2 = ExampleClass()
loss = Loss((sim1, sim2))

def forward():
    sim1.test()
    sim2.test()
    loss.compute_loss()

with ti.Tape(loss.loss):
    forward()

print(loss.loss[None])
print(sim1.x.grad)

```

**Log/Screenshots**
```
$ python sample_code.py
[Taichi] version 0.8.7, llvm 10.0.0, commit 88d81df6, linux, python 3.6.9
[Taichi] Starting on arch=x64
0.0
[nan  0.]
```
If you have local commits (e.g. compile fixes before you reproduce the bug), please make sure you first make a PR to fix the build errors and then report the bug.

**Comment by bobcao3** [comment]
This seems like another `0.0 / 0.0` case? `norm({x, y}) = sqrt(x * x + y * y)`, then `dnorm/dy = y / sqrt(x * x + y * y)`. When `x` and `y` are both zero and if the expression is evaluated normally under IEEE 754, the result is `NaN`. Maybe add an special case in the diff function?

**Comment by erizmr** [comment]
Sorry for the late reply. Let me have a look.

**Comment by erizmr** [comment]
Thanks for reporting this @kracon7 . As @bobcao3 mentioned, it is a ```0.0 / 0.0``` case. I think it is more like a case that users should try to avoid (similar to avoiding dividing by ```0``` ) when writing code. In addition, the autodiff system in Taichi do source code transformation in compile time. It does not have the information from runtime i.e. hard to handle these edge input cases.


BTW, pytorch also gives ```nan``` in similar case with all inputs zero.
```
import torch
a = torch.tensor(0.0, requires_grad=True)
b = torch.tensor(0.0, requires_grad=True)
d = torch.tensor(0.0, requires_grad=True)
c = d * torch.sqrt(a + b)
c.backward()
print(a.grad) # tensor(nan)
```

**Comment by bobcao3** [comment]
It seems we are consistent with `pytorch`, closing this issue for now. If you have more questions, feel free to re-open this issue.
